WcaOnRails/app/controllers/api/v0/competitions_controller.rb

```diff
@@ -1,11 +1,62 @@
 # frozen_string_literal: true
 class Api::V0::CompetitionsController < Api::V0::ApiController
-  def show
-    competition = Competition.where(id: params[:id], showAtAll: true).first
+  def competition_from_params
+    id = params[:competition_id] || params[:id]
+    competition = Competition.visible.where(id: id).first
     unless competition
-      render json: { error: "Competition with id #{params[:id]} not found" }, status: 404
-      return
+      render json: { error: "Competition with id #{id} not found" }, status: 404
+      return nil
     end
+    competition
+  end
+
+  def index
+    competitions = Competition.search(params[:q], params: params, current_resource_owner: current_resource_owner)
+    competitions = competitions.includes(:delegates).includes(:organizers)
+
+    paginate json: competitions
+  end
+
+  def show
+    competition = competition_from_params
+    return unless competition
+
     render json: competition
   end
+
+  def can_manage_competition?(competition)
+    unless current_resource_owner
+      render json: { error: "Not authenticated" }, status: 401
+      return false
+    end
+
+    unless doorkeeper_token.scopes.exists?("manage_competitions") && current_resource_owner.can_manage_competition?(competition)
+      render json: { error: "Not authorized to manage competition" }, status: 403
+      return false
+    end
+
+    true
+  end
+
+  def big_json
+    competition = competition_from_params
+    return unless competition
+    return unless can_manage_competition?(competition)
```

Comment:
This code actually just got reworked a bit, as I realized the original implementation did not allow you to work with hidden competitions.

I personally am not the biggest fan of `before_action` because I find they separate permission/validation checking logic from the actions and can make the actions hard to read in isolation.